### PR TITLE
Wait for with query

### PR DESCRIPTION
**Pull request by SimonRichardson**
## Description of change

This PR is the initial idea for moving `wait-for` semantics from `juju status` and
instead using the AllWatcher API. The initial idea is to allow the user to specify
via simple logical operators in a query form passed to the command a way to
see if a model or application has reached a given state.

The type of queries you can perform here are simple in nature until we pad this
out more. There are other types we also want to model, things like units, relations,
remote applications, and charms, just to name a few! For now, the PR is simple
and the code can be improved upon in the future.

----

The code uses a recursive descent Pratt parser for the query compiler (bringing back
my compiler lessons from my masters :scream: ). The query language supports
a limited set of operators and logical operators and no statements (essentially
only expressions). There is currently no number operators, I'll add them at a
later date. Grouping of expressions and precedence has been added so
correct resolution should be working.

The query evaluator/interpreter doesn't know about the deltas of a given entity
info, instead, we introduce scopes for finding given identifier values. For example,
a `ModelUpdate.Name` will be accessed via `name`. Aliases to more complex
objects also exist to make scope resolution much simpler. `status` instead
of `Status.Current`.

Equality checking is handled via `==`, or `reflect.DeepEqual`, otherwise it
will attempt to find out the underlying shadowed type and use that instead
(`type Life string` is a prime example).

----

The application code is more complicated because of the way the AllWatcher
forces you to handle the application status and I don't know of a better way to
fix that upstream.

----

The commands are split into entity types, so you can query per entity. In the future
it would be great to add `juju wait-for` with a goal state so that you don't have to
query everything at once.

----

Somethings that are possible (see Q&A)

```sh
$ juju wait-for application redis --query='name == "redis"'
$ juju wait-for application redis --query='(name == "redis" && life != "dead")'
$ juju wait-for application redis --query='exposed == false || (status == "unknown" || status == "active") && life != "dead"'
$ juju wait-for application redis --query='name == "redis"' --timeout=1m
```

## QA steps

```sh
$ juju bootstrap lxd test
$ juju deploy redis
$ juju wait-for application redis --query='status=="unknown"'
```

### CODE REVIEW

File: cmd/plugins/juju-wait-for/application.go [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+		Args:    "[<name>]",
+		Purpose: "wait for an application to reach a goal state",
+		Doc:     applicationCommandDoc,
+	})
+}
+
+// SetFlags implements Command.SetFlags.
+func (c *applicationCommand) SetFlags(f *gnuflag.FlagSet) {
+	c.waitForCommandBase.SetFlags(f)
+	f.StringVar(&c.query, "query", `life=="alive" && status=="active"`, "query the goal state")
```
Comment (manadart) [inline cmd/plugins/juju-wait-for/application.go:72]: Nice default.

File: cmd/plugins/juju-wait-for/application.go [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+
+	if !c.found {
+		logger.Infof("application %q not found, waiting...", name)
+		return false
+	}
+
+	var logOutput bool
+	currentStatus := c.appInfo.Status.Current
+
+	// If the application is unset, the derive it from the units.
```
Comment (manadart) [inline cmd/plugins/juju-wait-for/application.go:133]: "then derive..."

File: cmd/plugins/juju-wait-for/model.go [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+		Args:    "[<name>]",
+		Purpose: "wait for an application to reach a goal state",
+		Doc:     modelCommandDoc,
+	})
+}
+
+// SetFlags implements Command.SetFlags.
+func (c *modelCommand) SetFlags(f *gnuflag.FlagSet) {
+	c.waitForCommandBase.SetFlags(f)
+	f.StringVar(&c.query, "query", "life==available", "query the goal state")
```
Comment (manadart) [inline cmd/plugins/juju-wait-for/model.go:71]: This should be `status==available`

Comment (SimonRichardson) [reply cmd/plugins/juju-wait-for/model.go:71]: Damn, you're right!

**Comment by SimonRichardson** [comment]
$$merge$$

**Pull request merged**